clang/runtime/dpct-rt/include/dpl_extras/memory.h

```diff
@@ -670,6 +670,237 @@ template <typename T> class device_iterator : public device_pointer<T> {
+    return p;
+  }
+  tagged_pointer &operator++() {
+    ++this->m_ptr;
+    return *this;
+  }
+  tagged_pointer &operator--() {
+    --this->m_ptr;
+    return *this;
+  }
+  difference_type operator-(const tagged_pointer &it) const {
+    return this->m_ptr - it.m_ptr;
+  }
+  tagged_pointer &operator+=(difference_type forward) {
+    this->m_ptr = this->m_ptr + forward;
+    return *this;
+  }
+  tagged_pointer &operator-=(difference_type backward) {
+    this->m_ptr = this->m_ptr - backward;
+    return *this;
+  }
+};
+
+// Void specialization for tagged pointers. Iterator traits are not provided but
+// conversion to other non-void tagged pointers is allowed. Pointer arithmetic
+// is disallowed with this specialization.
+template <typename Tag>
+class tagged_pointer<Tag, void> : public tagged_pointer_base<Tag, void> {
+  using super_t = tagged_pointer_base<Tag, void>;
+
+public:
+  using difference_type = typename super_t::difference_type;
+  using pointer = typename super_t::pointer;
+  tagged_pointer() : super_t() {}
+  tagged_pointer(pointer ptr) : super_t(ptr) {}
+  // Enable tagged void pointer to convert to all other raw pointer types.
+  template <typename OtherPtr> operator OtherPtr *() const {
+    return static_cast<OtherPtr *>(this->m_ptr);
+  }
+};
+
+namespace internal {
+
+// Internal utility to return raw pointer to allocated memory. Note that host
+// allocations are not device accessible (not pinned).
+template <typename PolicyOrTag>
+void *malloc_base(PolicyOrTag &&policy_or_tag, const ::std::size_t num_bytes) {
+  using tag = policy_or_tag_to_tag<PolicyOrTag>;
+  if constexpr (::std::is_same_v<typename tag::type, host_sys_tag>) {
+    return ::std::malloc(num_bytes);
+  } else {
+    sycl::queue q;
+    // Grab the associated queue if a device policy is provided. Otherwise, use
+    // default constructed.
+    if constexpr (tag::is_policy_v) {
+      q = policy_or_tag.queue();
+    } else {
+      q = get_default_queue();
+    }
+    return sycl::malloc_shared(num_bytes, q);
+  }
+}
+
+} // namespace internal
+
+template <typename PolicyOrTag>
+auto malloc(PolicyOrTag &&policy_or_tag, const ::std::size_t num_bytes) {
+  return tagged_pointer<internal::policy_or_tag_to_tag_t<PolicyOrTag>, void>(
+      internal::malloc_base(::std::forward<PolicyOrTag>(policy_or_tag),
+                            num_bytes));
+}
+
+template <typename T, typename PolicyOrTag>
+auto malloc(PolicyOrTag &&policy_or_tag, const ::std::size_t num_elements) {
+  return tagged_pointer<internal::policy_or_tag_to_tag_t<PolicyOrTag>, T>(
+      static_cast<T *>(
+          internal::malloc_base(::std::forward<PolicyOrTag>(policy_or_tag),
+                                num_elements * sizeof(T))));
+}
+
+template <typename PolicyOrTag, typename Pointer>
+void free(PolicyOrTag &&policy_or_tag, Pointer ptr) {
+  using tag = internal::policy_or_tag_to_tag<PolicyOrTag>;
+  if constexpr (::std::is_same_v<typename tag::type, host_sys_tag>) {
```

Comment:
Does it make sense to abstract this out to something like:
`dpct::is_host_policy_or_tag_v<PolicyOrTag>`?

If you do this, you'd have to adjust `tag::is_policy_v`  a bit as well, probably to use `oneapi::dpl::execution::is_execution_policy_v`

Similar above.

Reply:
I have added `dpct::is_host_policy_or_tag_v<PolicyOrTag>` and switched `tag::is_policy_v` to just use `oneapi::dpl::execution::is_execution_policy_v` on the decayed `PolicyOrType`.